quesma/optimize/split_time_range.go

```diff
@@ -0,0 +1,274 @@
+
+import (
+	"fmt"
+	"quesma/logger"
+	"quesma/model"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// Context: We noticed that for some schemas (that don't ORDER BY time), the "Discover" view in Kibana
+// over long time ranges can be very slow, even though it only shows 500 results. Changing the time range
+// to a shorter one can make the query faster. (See this issue in ClickHouse for a similar example:
+// https://github.com/ClickHouse/ClickHouse/issues/69315)
+//
+// This optimization therefore splits the time range into parts: a short time range, on which we bet that the query
+// will be fast (and still return LIMIT many results) and a long time range, which will be used to get the rest of the
+// results (in case the short time range didn't return enough results).
+type splitTimeRange struct{}
+
+var defaultShorterTimeRangesMinutes = []int64{15}
+
+type timeRangeLimit struct {
+	value    int64
+	funcName string // for example fromUnixTimestamp64Milli
+}
+
+type timeRange struct {
+	columnName string
+	lowerLimit timeRangeLimit
+	upperLimit timeRangeLimit
+	direction  model.OrderByDirection
+}
+
+func (s splitTimeRange) validateSelectedColumns(columns []model.Expr) bool {
+	// The main purpose is to disallow window functions for which this optimization might be hard to reason about
+	// (and could be invalid). The allowed Expr types are whitelisted here rather than blacklisted (window functions)
+	// to be less error-prone in the future.
+	for i, column := range columns {
+		if _, ok := column.(model.ColumnRef); ok {
+			continue
+		}
+		if _, ok := column.(model.LiteralExpr); ok {
+			continue
+		}
+		if aliasedExpr, ok := column.(model.AliasedExpr); ok && s.validateSelectedColumns([]model.Expr{aliasedExpr.Expr}) {
+			continue
+		}
+		if functionExpr, ok := column.(model.FunctionExpr); ok && s.validateSelectedColumns(functionExpr.Args) {
+			continue
+		}
+
+		logger.Debug().Msgf("Query not eligible for time range optimization: column at index %d is an unsupported %T", i, column)
+		return false
+	}
+	return true
+}
+
+func (s splitTimeRange) findOrderByColumn(selectCommand *model.SelectCommand) (string, model.OrderByDirection, bool) {
+	if len(selectCommand.OrderBy) != 1 {
+		logger.Debug().Msg("Query not eligible for time range optimization: ORDER BY longer than 1")
+		return "", model.DefaultOrder, false
+	}
+
+	if orderByColumn, ok := selectCommand.OrderBy[0].Expr.(model.ColumnRef); ok {
+		return orderByColumn.ColumnName, selectCommand.OrderBy[0].Direction, true
+	}
+
+	logger.Debug().Msg("Query not eligible for time range optimization: ORDER BY not a column reference")
+	return "", model.DefaultOrder, false
+}
+
+func (s splitTimeRange) checkAndFindTimeLimits(selectCommand *model.SelectCommand, orderByColumnName string) (*timeRangeLimit, *timeRangeLimit) {
+	var lowerLimit, upperLimit *timeRangeLimit
+
+	visitor := model.NewBaseVisitor()
+	visitor.OverrideVisitInfix = func(b *model.BaseExprVisitor, e model.InfixExpr) interface{} {
```

Comment:
Unfortunately, this does not seem to be correct. We should only check where part of the statement and top-level conditions are separated by AND. This does not support also case when there are multiple filtering of timestamp.